File: py/tuish_textual/widgets/license_gate.py
"""LicenseGate widget for conditional rendering based on license status."""

from __future__ import annotations

from collections.abc import Sequence
from typing import TYPE_CHECKING

from textual import work
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import LoadingIndicator, Static

from tuish_textual.messages import LicenseChanged, LicenseCheckComplete, LicenseCheckStarted

if TYPE_CHECKING:
    from textual.app import ComposeResult
    from tuish import Tuish
    from tuish.models import LicenseCheckResult


class LicenseGate(Widget):
    """Conditionally renders children based on license status or feature availability.

    This widget checks the license on mount and renders different content based
    on the result:
    - While checking: shows loading content
    - When access granted: shows children
    - When access denied: shows fallback content

    Example:
        ```python
        # Gate by feature
        yield LicenseGate(
            tuish=tuish,
            feature="pro",
            fallback=UpgradePrompt(),
            children=[ProFeatureWidget()],
        )

        # Gate by any valid license
        yield LicenseGate(
            tuish=tuish,
            require_license=True,
            fallback=PurchaseFlow(tuish=tuish),
            children=[PaidContent()],
        )
        ```

    Attributes:
        feature: Specific feature flag to check for access. If None, just checks
            for a valid license (when require_license=True).
        require_license: Whether any valid license is required. Only used when
            feature is None. Defaults to True.
        is_loading: Whether a license check is in progress.
        has_access: Whether access has been granted based on license check.
    """

    DEFAULT_CSS = """
    LicenseGate {
        width: 100%;
        height: auto;
    }

    LicenseGate > .license-gate-loading {
        width: 100%;
        height: 3;
        content-align: center middle;
    }

    LicenseGate > .license-gate-message {
        width: 100%;
        text-align: center;
        color: $text-muted;
    }
    """

    is_loading: reactive[bool] = reactive(True, recompose=True)
    """Whether a license check is currently in progress."""

    has_access: reactive[bool] = reactive(False, recompose=True)
    """Whether access has been granted based on the license check."""

    def __init__(
        self,
        tuish: Tuish,
        *,
        feature: str | None = None,
        require_license: bool = True,
        children: Sequence[Widget] | None = None,
        fallback: Widget | Sequence[Widget] | None = None,
        loading: Widget | None = None,
        name: str | None = None,
        id: str | None = None,
        classes: str | None = None,
        disabled: bool = False,
    ) -> None:
        """Initialize the LicenseGate widget.

        Args:
            tuish: Tuish SDK instance for license checking.
            feature: Specific feature flag to check for access. If provided,
                access is granted only if the license includes this feature.
            require_license: Whether any valid license is required when no
                feature is specified. Defaults to True.
            children: Widgets to display when access is granted.
            fallback: Widget(s) to display when access is denied.
            loading: Widget to display while checking license.
            name: The name of the widget.
            id: The ID of the widget in the DOM.
            classes: Space-separated CSS class names.
            disabled: Whether the widget is disabled.
        """
        super().__init__(name=name, id=id, classes=classes, disabled=disabled)
        self._tuish = tuish
        self._feature = feature
        self._require_license = require_license

        # Store content widgets
        self._children: list[Widget] = list(children) if children else []
        self._fallback: list[Widget] = (
            [fallback] if isinstance(fallback, Widget) else list(fallback) if fallback else []
        )
        self._loading_widget = loading

        # Store the last license check result
        self._license_result: LicenseCheckResult | None = None

    @property
    def feature(self) -> str | None:
        """The feature flag to check for access."""
        return self._feature

    @property
    def require_license(self) -> bool:
        """Whether any valid license is required."""
        return self._require_license

    @property
    def license_result(self) -> LicenseCheckResult | None:
        """The result of the last license check."""
        return self._license_result

    def compose(self) -> ComposeResult:
        """Compose the widget content based on current state."""
        if self.is_loading:
            if self._loading_widget is not None:
                yield self._loading_widget
            else:
                yield LoadingIndicator(classes="license-gate-loading")
                yield Static("Checking license...", classes="license-gate-message")
        elif self.has_access:
            yield from self._children
        else:
            if self._fallback:
                yield from self._fallback
            else:
                yield Static("Access denied", classes="license-gate-message")

    def on_mount(self) -> None:
        """Start license check when widget is mounted."""
        self._check_license()

    @work(thread=True)
    def _check_license(self) -> None:
        """Check license in a background thread.

        Posts LicenseCheckStarted before checking and LicenseCheckComplete
        when done. Also posts LicenseChanged to notify parent widgets.
        """
        # Post that we're starting the check
        self.post_message(LicenseCheckStarted())

        # Perform the license check (this can be slow, runs in thread)
        result = self._tuish.check_license()
        self._license_result = result

        # Determine access based on feature or license requirement
        has_access = self._determine_access(result)

        # Update reactive properties (must be done on main thread via call_from_thread)
        self.app.call_from_thread(self._update_state, has_access, result)

    def _update_state(self, has_access: bool, result: LicenseCheckResult) -> None:
        """Update widget state after license check completes.

        Args:
            has_access: Whether access should be granted.
            result: The license check result.
        """
        self.has_access = has_access
        self.is_loading = False

        # Post completion messages
        self.post_message(LicenseCheckComplete(result=result))
        self.post_message(
            LicenseChanged(
                license=result.license,
                valid=result.valid,
                offline_verified=result.offline_verified,
            )
        )

    def _determine_access(self, result: LicenseCheckResult) -> bool:
        """Determine if access should be granted based on the license check result.

        Args:
            result: The license check result.

        Returns:
            True if access should be granted, False otherwise.
        """
        if self._feature is not None:
            # Feature-based gating: check if license has the specific feature
            if not result.valid or result.license is None:
                return False
            return self._feature in result.license.features

        if self._require_license:
            # Any valid license required
            return result.valid

        # No gating specified, allow access
        return True

    def recheck_license(self) -> None:
        """Trigger a re-check of the license.

        Call this method to refresh the license status, for example after
        a purchase is completed.
        """
        self.is_loading = True
        self.has_access = False
        self._check_license()

File: go/tui/license_gate.go
package tui

import (
	tea "github.com/charmbracelet/bubbletea"
	tuish "github.com/tuishdotdev/tuish/go"
)

// LicenseGateConfig contains configuration for the LicenseGate component.
type LicenseGateConfig struct {
	// Feature is the specific feature required for access.
	// If empty, RequireLicense is checked instead.
	Feature string

	// RequireLicense requires any valid license when Feature is not set.
	RequireLicense bool

	// Styles allows custom styling.
	Styles *Styles
}

// LicenseGate conditionally renders content based on license status.
// It wraps a child model and only allows access when licensing requirements are met.
type LicenseGate struct {
	sdk      *tuish.SDK
	config   LicenseGateConfig
	styles   Styles
	child    tea.Model
	fallback tea.Model
	loading  tea.Model

	result     *tuish.LicenseCheckResult
	isLoading  bool
	hasAccess  bool
	err        error
}

// NewLicenseGate creates a new LicenseGate that wraps a child model.
//
// The child model is displayed when access is granted.
// The fallback model is displayed when access is denied (optional).
// If no fallback is provided, a default "access denied" message is shown.
//
// Example:
//
//	gate := tui.NewLicenseGate(sdk, myApp, tui.LicenseGateConfig{
//		Feature: "pro",
//	})
//	gate.SetFallback(purchaseFlow)
func NewLicenseGate(sdk *tuish.SDK, child tea.Model, config ...LicenseGateConfig) *LicenseGate {
	cfg := LicenseGateConfig{}
	if len(config) > 0 {
		cfg = config[0]
	}

	styles := DefaultStyles()
	if cfg.Styles != nil {
		styles = *cfg.Styles
	}

	return &LicenseGate{
		sdk:       sdk,
		config:    cfg,
		styles:    styles,
		child:     child,
		isLoading: true,
	}
}

// SetFallback sets the model to display when access is denied.
func (m *LicenseGate) SetFallback(fallback tea.Model) *LicenseGate {
	m.fallback = fallback
	return m
}

// SetLoading sets a custom loading model.
func (m *LicenseGate) SetLoading(loading tea.Model) *LicenseGate {
	m.loading = loading
	return m
}

// Init initializes the LicenseGate by checking the license.
func (m *LicenseGate) Init() tea.Cmd {
	return m.checkLicense
}

// Update handles messages for the LicenseGate.
func (m *LicenseGate) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case LicenseCheckedMsg:
		m.isLoading = false
		if msg.Error != nil {
			m.err = msg.Error
			m.hasAccess = false
		} else {
			m.result = msg.Result
			m.hasAccess = m.checkAccess(msg.Result)
		}

		// Initialize the appropriate child model
		if m.hasAccess {
			return m, m.child.Init()
		}
		if m.fallback != nil {
			return m, m.fallback.Init()
		}
		return m, nil

	case LicenseStoredMsg:
		// Re-check license after storing
		if msg.Error == nil {
			m.isLoading = true
			return m, m.checkLicense
		}

	case LicenseRefreshedMsg:
		if msg.Error == nil && msg.Result != nil {
			m.result = msg.Result
			m.hasAccess = m.checkAccess(msg.Result)
		}
	}

	// Pass messages to the active child model
	if m.isLoading {
		if m.loading != nil {
			var cmd tea.Cmd
			m.loading, cmd = m.loading.Update(msg)
			return m, cmd
		}
		return m, nil
	}

	if m.hasAccess {
		var cmd tea.Cmd
		m.child, cmd = m.child.Update(msg)
		return m, cmd
	}

	if m.fallback != nil {
		var cmd tea.Cmd
		m.fallback, cmd = m.fallback.Update(msg)
		return m, cmd
	}

	return m, nil
}

// View renders the LicenseGate.
func (m *LicenseGate) View() string {
	if m.isLoading {
		if m.loading != nil {
			return m.loading.View()
		}
		return m.styles.Muted.Render("Checking license...")
	}

	if m.hasAccess {
		return m.child.View()
	}

	if m.fallback != nil {
		return m.fallback.View()
	}

	return m.renderAccessDenied()
}

func (m *LicenseGate) renderAccessDenied() string {
	if m.config.Feature != "" {
		return m.styles.BoxWarning.Render(
			m.styles.Warning.Render(Lock+" Feature Required") + "\n\n" +
				m.styles.Body.Render("The \""+m.config.Feature+"\" feature requires a valid license.") + "\n" +
				m.styles.Muted.Render("Please upgrade to access this feature."),
		)
	}

	return m.styles.BoxWarning.Render(
		m.styles.Warning.Render(Lock+" License Required") + "\n\n" +
			m.styles.Body.Render("A valid license is required to access this application.") + "\n" +
			m.styles.Muted.Render("Please purchase a license to continue."),
	)
}

func (m *LicenseGate) checkAccess(result *tuish.LicenseCheckResult) bool {
	if m.config.Feature != "" {
		// Feature-based gating
		return m.hasFeature(result, m.config.Feature)
	}

	if m.config.RequireLicense {
		// Any valid license required
		return result.Valid
	}

	// No gating specified, allow access
	return true
}

func (m *LicenseGate) hasFeature(result *tuish.LicenseCheckResult, feature string) bool {
	if result == nil || result.License == nil {
		return false
	}

	for _, f := range result.License.Features {
		if f == feature {
			return true
		}
	}
	return false
}

func (m *LicenseGate) checkLicense() tea.Msg {
	result, err := m.sdk.CheckLicense(nil)
	return LicenseCheckedMsg{Result: result, Error: err}
}

// HasAccess returns whether access is currently granted.
func (m *LicenseGate) HasAccess() bool {
	return m.hasAccess
}

// IsLoading returns whether the component is loading.
func (m *LicenseGate) IsLoading() bool {
	return m.isLoading
}

// Result returns the current license check result.
func (m *LicenseGate) Result() *tuish.LicenseCheckResult {
	return m.result
}

// Child returns the wrapped child model.
func (m *LicenseGate) Child() tea.Model {
	return m.child
}

// Refresh triggers a license refresh.
func (m *LicenseGate) Refresh() tea.Cmd {
	m.isLoading = true
	return m.checkLicense
}

// SimpleLicenseGate provides a simpler interface for gating without a full Bubble Tea model.
// It checks the license synchronously and returns access status.
type SimpleLicenseGate struct {
	sdk     *tuish.SDK
	feature string
}

// NewSimpleLicenseGate creates a simple license gate for synchronous checks.
func NewSimpleLicenseGate(sdk *tuish.SDK, feature ...string) *SimpleLicenseGate {
	f := ""
	if len(feature) > 0 {
		f = feature[0]
	}
	return &SimpleLicenseGate{sdk: sdk, feature: f}
}

// Check performs a synchronous license check and returns access status.
func (g *SimpleLicenseGate) Check() (hasAccess bool, result *tuish.LicenseCheckResult, err error) {
	result, err = g.sdk.CheckLicense(nil)
	if err != nil {
		return false, nil, err
	}

	if g.feature != "" {
		// Check for specific feature
		if result.License != nil {
			for _, f := range result.License.Features {
				if f == g.feature {
					return true, result, nil
				}
			}
		}
		return false, result, nil
	}

	// Just check for valid license
	return result.Valid, result, nil
}

// HasFeature checks if the current license has a specific feature.
func HasFeature(sdk *tuish.SDK, feature string) bool {
	result, err := sdk.CheckLicense(nil)
	if err != nil || result == nil || result.License == nil {
		return false
	}

	for _, f := range result.License.Features {
		if f == feature {
			return true
		}
	}
	return false
}

// IsLicensed checks if the current license is valid.
func IsLicensed(sdk *tuish.SDK) bool {
	result, err := sdk.CheckLicense(nil)
	if err != nil {
		return false
	}
	return result.Valid
}
